### PR TITLE
Separate preparation timeouts for PVF prechecking and execution

### PR DESCRIPTION
**Pull request by m-cat**
# PULL REQUEST

## Overview

Per the linked issue, we make the required changes so that preparation for
execution is more lenient (by a factor of 3) than preparation for prechecking.

We [add a compilation_timeout parameter for PVF preparation job](https://github.com/paritytech/polkadot/commit/ee0abe26cc1b81c961ee1b78925d64795c305bd6) 
and also split the `COMPILATION_TIMEOUT` constant into two new consts.

### Todo

- [ ] ~~What kind of tests should be added?~~

### CODE REVIEW

**Comment by m-cat** [comment]
Not sure how to label this for release, i.e. what is worthy of release notes?

**Comment by rphmeier** [comment]
@m-cat - the default should be 'silent' but major features, API changes, DB locations, and other things that impact the process of running a validator should be present in release notes.

**Review by eskimor — APPROVED** [review]
Excellent! Really nice work! One thing, with regards to leniency, for execution we have a factor of 6, having the same factor here seems to make sense and would make it a bit "safer".

File: node/core/pvf/src/prepare/queue.rs
Comment (pepyakin) [inline node/core/pvf/src/prepare/queue.rs:83]: Nit: I think here (and elsewhere) it's more correct to say `preparation`. The worker will perform preparation, which is the combination of prevalidation and compilation (production of the compiled artifact).

File: node/core/pvf/src/host.rs
Comment (pepyakin) [inline node/core/pvf/src/host.rs:42]: Nit: those docs distinguish between pre-check preparation worker and execute preparation worker. I am not sure if that's the right way of thinking about that. After all it's the very same worker that does the same the job. It does not even have any parametrization.

**Comment by eskimor** [comment]
With regards to tests, for this particular change there is hardly something we could test. On

> Excellent! Really nice work! One thing, with regards to leniency, for execution we have a factor of 6, having the same factor here seems to make sense and would make it a bit "safer".

Ok, actually the situation is a bit different. For backing/approval timeouts we only have like 2-3 validators with the short timeout, here we have a super majority - so a smaller lenience can be justified - still node load can vary over time, also the validator set may change over time ... better safe than sorry.

File: node/core/pvf/src/host.rs
Comment (pepyakin) [inline node/core/pvf/src/host.rs:49]: Nit: It would be great if there were a doc line explaining the relationship between the two. Perhaps, moving them into a module (inline or separate file) and in the module doc explaining the stuff we discussed in DMs?

File: node/core/pvf/src/metrics.rs
Comment (pepyakin) [inline node/core/pvf/src/metrics.rs:170]: Nit: Do you think that's a good resolution for this metric? IOW, ask the question, looking at a metrics dashboard do you think it's possible that you would think for yourself "I wish there were more buckets available that are higher/lesser than 180"? If you are inclined to say yes just plop more bands. It's very cold code.

File: roadmap/implementers-guide/src/pvf-prechecking.md
Comment (pepyakin) [inline roadmap/implementers-guide/src/pvf-prechecking.md:25]: Nit: This book already has a glossary. Do you think it would be better to move those there? Here, we can leave a note saying that this is a loaded document with terms refer to the glossary.

Alternatively (or better, additionally) we could embed those into the text as explainations. I think as a bonus this would allow us to structure the explaination hierarchically, IMO better. Something like the following abstract:

1. In order to make the PVF usable for candidate validation it has to be registered on-chain
1. As part of the registration process, it has to go through pre-checking.
1. Pre-checking is a game of attempting preparation and reporting the results back on-chain.
1. We define preparation as a process that: validates the consistency of the wasm binary (aka prevalidation) and the compilation of the wasm module into machine code (refered to as artifact).
1. Besides pre-checking, preparation can also be triggered by execution, since compiled artifact is needed for the execution

File: node/core/pvf/src/host.rs
Comment (pepyakin) [inline node/core/pvf/src/host.rs:49]: Nit: I wonder if this would be better named as `lazy` or `lenient`. After all we use it for the heads up signal which also requires a more permissive timeout.

File: node/core/pvf/src/host.rs
Comment (pepyakin) [inline node/core/pvf/src/host.rs:432]: Nit: ... prechecking requests

File: node/core/pvf/src/host.rs
Comment (pepyakin) [inline node/core/pvf/src/host.rs:514]: lol, finally this is fixed 🎉

**Review by pepyakin — APPROVED** [review]
Good job. Don't get discouraged by the number of comments. They are all nits, which (at least for my reviews) means they are not blocking the merge and are not necessary to address. 

Organizational note: I personally prefer landing code as soon as it's ready. If you see a part that is not ready and you wanna address before the merge, I am up to split it into another PR. That will give us a platform for discussion the changes with more focus. It's also possible to land and as-is and send the fixes with a follow-up PR (as long as everything is linked)

**Comment by m-cat** [comment]
Thanks for the reviews! 👍 Just stuck on a couple of CI checks:

`Check reviews`: Looks like I need a couple reviews from the `ci` or `release-engineering` teams.
`zombienet`: I'm trying to restart this one on GitLab, but getting `This job could not start because it could not retrieve the needed artifacts: publish-polkadot-debug-image`.

**Comment by m-cat** [comment]
> still node load can vary over time, also the validator set may change over time ... better safe than sorry.

@eskimor Sounds good, I'll address it in the followup PR.

> Good job. Don't get discouraged by the number of comments. They are all nits, which (at least for my reviews) means they are not blocking the merge and are not necessary to address.

All good. I was expecting more comments than that for my first PR! 

I agree about addressing the nits in a followup PR.

File: node/core/pvf/src/metrics.rs
Comment (m-cat) [reply node/core/pvf/src/metrics.rs:170]: What do you mean by cold code?

**Comment by ordian** [comment]
jFYI: `continuous-integration/gitlab-zombienet-tests-parachains-disputes` is expected to fail for now on master, it's marked as `{"build_allow_failure":true}`
should be fixed in #6142

**Review by sergejparity — APPROVED** [review]

**Review by alvicsam — APPROVED** [review]

**Review by slumber — APPROVED** [review]
Agreed with @pepyakin comments, other than that well done

Note: once pvf is queued for preparation with some timeout, any subsequent request would discard `compilation_timeout` parameter and simply enqueue `response_receiver` (the way it's implemented now)

This is OK because we can't receive execute request for unprepared code until it's enacted once prechecking process concludes, and the code that was already prechecked should pass this process with a greater timeout.

However, this is an external guarantee so wanted to make sure you keep it in mind.

File: node/core/pvf/src/metrics.rs
Comment (pepyakin) [reply node/core/pvf/src/metrics.rs:170]: generally, cold code is the code that is not frequently called. Here, I meant that there are no performance reasons to save on the bands. Assuming more bands, less performance. I did not even think about this too much since it's how many preparations per second can we reasonably do in the worst case? 100? So with the performance of the node argument being irrelevant, and by extension memory as well. I don't see other arguments against it.

**Pull request merged**
